abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,919 @@
+Called for each HTTP/2 METADATA frame received from the upstream.  Metadata can be retrieved using
+`proxy_get_map` and/or `proxy_get_map_value`.
+
+
+## HTTP calls
+
+### `proxy_on_http_call_response`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) callout_id`
+  - `i32 (size_t) num_headers`
+  - `i32 (size_t) body_size`
+  - `i32 (size_t) num_trailers`
+* returns:
+  - none
+
+Called when the response to the HTTP call (`callout_id`) is received.
+
+
+## gRPC callouts
+
+### `proxy_on_grpc_call_response_header_metadata`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) callout_id`
+  - `i32 (size_t) num_elemets`
+* returns:
+  - none
+
+Called when header metadata in the response to the gRPC call (`callout_id`) is received.
+
+
+### `proxy_on_grpc_call_response_message`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) callout_id`
+  - `i32 (size_t) message_size`
+* returns:
+  - none
+
+Called when the response to the gRPC call (`callout_id`) is received.
+
+
+### `proxy_on_grpc_call_response_trailer_metadata`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) callout_id`
+  - `i32 (size_t) num_elements`
+* returns:
+  - none
+
+Called when trailer metadata in the response to the gRPC call (`callout_id`) is received.
+
+
+### `proxy_on_grpc_call_close`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) callout_id`
+  - `i32 (uint32_t) status_code`
+* returns:
+  - none
+
+Called when the request to the gRPC call (`callout_id`) fails or if the gRPC connection is closed.
+
+
+## Shared queue
+
+### `proxy_on_queue_ready`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) queue_id`
+* returns:
+  - none
+
+Called when there is data available in the queue (`queue_id`).
+
+
+# Functions implemented in the host environment
+
+All functions implemented in the host environment return `proxy_result_t`, which indicates the
+status of the call (successful, invalid memory access, etc.), and the return values are written into
+memory pointers passed in as arguments (indicated by the `return_` prefix in the specification).
+
+
+## Logging
+
+### `proxy_log`
+
+* params:
+  - `i32 (proxy_log_level_t) log_level`
+  - `i32 (const char*) message_data`
+  - `i32 (size_t) message_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Log message (`message_data`, `message_size`) at the given `log_level`.
+
+
+### `proxy_get_current_time`
+
+* params:
+  - `i32 (uint64_t*) return_current_time_nanoseconds`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get current time (`return_current_time`).
+
+
+## Module lifecycle
+
+### `proxy_set_effective_context`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Change the effective context to `context_id`. This function is usually used to change the context
+after receiving `proxy_on_http_call_response`, `proxy_on_grpc_call_response` or
+`proxy_on_queue_ready`.
```

Comment:
is this needed because these callbacks are called in the context of the root context?